sycl/include/sycl/ext/oneapi/reduction.hpp

```diff
@@ -1381,36 +1390,11 @@ reduCGFuncImpl(handler &CGH, KernelType KernelFunc, const nd_range<Dims> &Range,
   });
 }
 
-template <typename KernelName, typename KernelType, int Dims, class Reduction>
-enable_if_t<Reduction::has_fast_atomics>
-reduCGFunc(handler &CGH, KernelType KernelFunc, const nd_range<Dims> &Range,
-           Reduction &Redu) {
-
-  size_t WGSize = Range.get_local_range().size();
-
-  // User's initialized read-write accessor is re-used here if
-  // initialize_to_identity is not set (i.e. if user's variable is initialized).
-  // Otherwise, a new buffer is initialized with identity value and a new
-  // read-write accessor to that buffer is created. That is done because
-  // atomic operations update some initialized memory.
-  // User's USM pointer is not re-used even when initialize_to_identity is not
-  // set because it does not worth the creation of an additional variant of
-  // a user's kernel for that case.
-  auto Out = Redu.getReadWriteAccessorToInitializedMem(CGH);
-
-  // If the work group size is not pow of 2, then the kernel runs some
-  // additional code and checks in it.
-  // If the reduction has fast reduce then the kernel does not care if the work
-  // group size is pow of 2 or not, assume true for such cases.
-  bool IsPow2WG = Reduction::has_fast_reduce || ((WGSize & (WGSize - 1)) == 0);
-  if (IsPow2WG)
-    reduCGFuncImpl<KernelName, KernelType, Dims, Reduction, true>(
-        CGH, KernelFunc, Range, Redu, Out);
-  else
-    reduCGFuncImpl<KernelName, KernelType, Dims, Reduction, false>(
-        CGH, KernelFunc, Range, Redu, Out);
-}
-
+namespace reduction {
+namespace main_krn {
+template <class KernelName> struct NDRangeBothFastReduceOnly;
```

Comment:
I assume this was supposed to be
```suggestion
template <class KernelName> struct NDRangeFastReduceOnly;
```